CMakeLists.txt

```diff
@@ -78,6 +81,13 @@ if (ENABLE_TESTING)
     enable_testing()
 endif()
 
+if (ENABLE_TESTING OR ENABLE_ONLY_SHARED)
```

Comment:
To my understanding,  you want to use shared libs for UTs and give an option for binaries,  right? 
If so, why not set ENABLE_ONLY_SHARED to be true directly when ENABLE_TESTING switch on.  So you check logic below will be much simpler. 
Correct me if i am wrong.